### PR TITLE
EX-2270: integrate clipper rfq

### PR DESCRIPTION
**Pull request by lehainam-dev**

### CODE REVIEW

File: pkg/liquidity-source/clipper/client/http.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		config: config,
+	}
+}
+
+func (c *httpClient) RFQ(ctx context.Context, params clipper.QuoteParams) (clipper.SignResponse, error) {
+	// 1. Call quote endpoint
+	req := c.client.R().SetContext(ctx).SetBody(params)
+
+	var quoteRes clipper.QuoteResponse
+	res, err := req.SetResult(&quoteRes).Post(quotePath)
```
Comment (NgoKimPhu) [inline pkg/liquidity-source/clipper/client/http.go:44]: if the response has status > 299, need to also `SetError(&quoteRes)` to read the body

File: pkg/liquidity-source/clipper/client/http.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	req = c.client.R().SetContext(ctx).SetBody(clipper.SignParams{
+		QuoteID:            quoteRes.ID,
+		DestinationAddress: params.DestinationAddress,
+		SenderAddress:      params.SenderAddress,
+		NativeInput:        false,
+		NativeOutput:       false,
+	})
+
+	var signRes clipper.SignResponse
+	res, err = req.SetResult(&signRes).Post(signPath)
```
Comment (NgoKimPhu) [inline pkg/liquidity-source/clipper/client/http.go:63]: same here

File: pkg/liquidity-source/clipper/client/http.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	req := c.client.R().SetContext(ctx).SetBody(params)
+
+	var quoteRes clipper.QuoteResponse
+	res, err := req.SetResult(&quoteRes).Post(quotePath)
+	if err != nil {
+		return clipper.SignResponse{}, err
+	}
+
+	if !res.IsSuccess() {
+		return clipper.SignResponse{}, ErrQuoteFailed
```
Comment (NgoKimPhu) [inline pkg/liquidity-source/clipper/client/http.go:50]: parsing and mapping the error is optional but we should still have some logs to know why it fails

File: pkg/liquidity-source/clipper/pool_simulator.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+			new(big.Float).Mul(M, inX),
+			pX,
+		),
+		pY,
+	)
+	outY.Mul(outY, bignumber.TenPowDecimals(assetOut.Decimals))
+
+	amountOut, _ := outY.Int(nil)
+
+	// TODO: Apply the final check FMV
```
Comment (NgoKimPhu) [inline pkg/liquidity-source/clipper/pool_simulator.go:141]: TODO?

File: pkg/liquidity-source/clipper/pool_simulator.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	M := new(big.Float).SetFloat64((basisPoint - pairInfo.FeeInBasisPoints) / basisPoint)
+
+	outY := new(big.Float).Quo(
+		new(big.Float).Mul(
+			new(big.Float).Mul(M, inX),
+			pX,
+		),
+		pY,
+	)
+	outY.Mul(outY, bignumber.TenPowDecimals(assetOut.Decimals))
```
Comment (NgoKimPhu) [inline pkg/liquidity-source/clipper/pool_simulator.go:137]: (just a comment) so many `new(big.Float)` 😭

File: pkg/liquidity-source/clipper/pool_simulator.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+		TokenAmountOut: &pool.TokenAmount{
+			Token:  assetOut.Address,
+			Amount: amountOut,
+		},
+		Fee: &pool.TokenAmount{
+			Token:  assetOut.Address,
+			Amount: bignumber.ZeroBI,
+		},
+
+		// TODO: Define gas
```
Comment (NgoKimPhu) [inline pkg/liquidity-source/clipper/pool_simulator.go:153]: gas gas gas

File: pkg/liquidity-source/clipper/client/http.go (hunk 2 above)
Comment (lehainam-dev) [reply pkg/liquidity-source/clipper/client/http.go]: Updated

File: pkg/liquidity-source/clipper/pool_simulator.go (hunk 4 above)
Comment (lehainam-dev) [reply pkg/liquidity-source/clipper/pool_simulator.go]: The FMV will always return true, I added a comment to explain.

File: pkg/liquidity-source/clipper/pool_simulator.go (hunk 5 above)
Comment (lehainam-dev) [reply pkg/liquidity-source/clipper/pool_simulator.go]: I move to the local variable `big.Float`, so the data can be allocated in stack instead.

File: pkg/liquidity-source/clipper/pool_simulator.go (hunk 6 above)
Comment (lehainam-dev) [reply pkg/liquidity-source/clipper/pool_simulator.go]: Updated

**Review by NgoKimPhu — APPROVED** [review]

**Pull request merged**